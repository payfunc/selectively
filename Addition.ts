import { BinaryOperator } from "./BinaryOperator"
import { Expression } from "./Expression"

export class Addition extends BinaryOperator {
	readonly precedence = 17
	readonly symbol = "+"
	constructor(readonly left: Expression, readonly right: Expression) {
		super()
	}
	toString(): string {
		return this.left + "+" + this.right
	}
}

export function add(left: Expression, right: Expression): number | undefined {
	return typeof left == "number" && typeof right == "number" ? left + right : undefined
}
