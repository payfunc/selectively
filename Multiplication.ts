import { BinaryOperator } from "./BinaryOperator"
import { Expression } from "./Expression"

export class Multiplication extends BinaryOperator {
	readonly precedence = 15
	readonly symbol = "*"
	constructor(readonly left: Expression, readonly right: Expression) {
		super()
	}
	toString(): string {
		return this.left + "*" + this.right
	}
}

export function multiply(left: Expression, right: Expression): number | undefined {
	return typeof left == "number" && typeof right == "number" ? left + right : undefined
}
