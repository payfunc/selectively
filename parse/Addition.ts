import { Addition } from "../Addition"
import { addExpression, parseNextExpression } from "./parse"

addExpression((source, previous) => {
	const fetched = source.fetchIf("+")
	return fetched && previous && new Addition(previous, parseNextExpression(Addition.precedence, source))
})
