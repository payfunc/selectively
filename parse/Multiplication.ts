import { Expression, getPrecedence } from "../Expression"
import { Multiplication } from "../Multiplication"
import { Value } from "../Value"
import { addExpression, parseNextExpression } from "./parse"

addExpression((source, previous) => {
	let result: Expression | undefined
	let right: Expression | undefined
	if (previous && source.fetchIf("*")) {
		if ((getPrecedence(source.peek(1)?.value) ?? 0) >= Multiplication.precedence) {
			right = parseNextExpression(Multiplication.precedence, source)
			result = new Multiplication(previous, right)
		} else
			result = new Multiplication(previous, parseNextExpression(new Value(0), source))
	}
	if (result)
		result = parseNextExpression(result, source)
	return result
})
