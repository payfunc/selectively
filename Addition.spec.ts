import { Addition } from "./Addition"
import { Multiplication } from "./Multiplication"

describe("selectively.Type", () => {
	it("Addition", () => {
		const adder = new Addition(5, 3)
		expect(adder).toMatchSnapshot()
	})
	it("Multiplication", () => {
		const adder = new Multiplication(5, 3)
		expect(adder).toMatchSnapshot()
	})
	it("Multiplication and Addition", () => {
		const adder = new Multiplication(new Addition(4, 5), 3)
		expect(adder).toMatchSnapshot()
	})
})
