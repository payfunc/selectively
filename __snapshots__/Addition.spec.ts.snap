// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectively.Type Addition 1`] = `
Addition {
  "left": 5,
  "precedence": 17,
  "right": 3,
  "symbol": "+",
}
`;

exports[`selectively.Type Multiplication 1`] = `
Multiplication {
  "left": 5,
  "precedence": 15,
  "right": 3,
  "symbol": "*",
}
`;

exports[`selectively.Type Multiplication and Addition 1`] = `
Multiplication {
  "left": Addition {
    "left": 4,
    "precedence": 17,
    "right": 5,
    "symbol": "+",
  },
  "precedence": 15,
  "right": 3,
  "symbol": "*",
}
`;
